docs/documentation/development/bindings/firmware-update.md

```diff
@@ -0,0 +1,118 @@
+otherwise false (e.g. if the thing is offline or if its status detail is already firmware updating).
+
+The thing returned by the `getThing()` operation must have set the `org.eclipse.smarthome.core.thing.Thing.PROPERTY_FIRMWARE_VERSION` property. If it is not set then the framework cannot determine the thing´s `org.eclipse.smarthome.core.thing.firmware.FirmwareStatus`.
+
+## Implementing the updateFirmware Operation
+
+Initially the implementation should change the thing status to `org.eclipse.smarthome.core.thing.ThingStatus.OFFLINE` and the thing status detail to `org.eclipse.smarthome.core.thing.ThingStatusDetail.FIRMWARE_UPDATING`. Afterwards if the handler would like to notify the framework of update progress information then the sequence of the firmware update progress steps has to be defined using the `defineSequence` operation of the given callback interface. Then by using the `next` operation of the callback interface the handler can indicate that the next progress step is going to be executed.
+
+```java 
+    public void updateFirmware(Firmware firmware, ProgressCallback progressCallback) {
+        // if thing is handled by a thing handler
+        updateStatus(ThingStatus.ONLINE, ThingStatusDetail.FIRMWARE_UPDATING);
+        
+        // define the sequence of the firmware update so that external consumers can listen for the progress
+        progressCallback.defineSequence(ProgressStep.DOWNLOADING, ProgressStep.TRANSFERRING, ProgressStep.UPDATING);
+        
+        // download / read firmware image
+        progressCallback.next();
+        ... read input stream from firmware.getContent();
+        
+        // transfer image to device
+        progressCallback.next();
+        ... transfer image to device
+        
+        // triggering the actual firmware update
+        progressCallback.next();
+        ... trigger update
+        
+        // here: send immediately the success information because it is not mandatory for this implementation to wait for the successful update of the device
+        progressCallback.success();
+        // or: if required to check whether the firmware update was successful or erroneous then wait or spawn a new thread
+    }
+```
+
+Finally the handler has to acknowledge in either case if the firmware update was successful or not. A successful firmware update is notified by the operation `progressCallback.success()` and an erroneous firmware update is notified by the operation `progressCallback.failed(String errorMessageKey, Object... arguments)`. In case of an erroneous firmware update the handler has to provide the error message key of the error message and optional arguments that are to be injected into the error message. The framework will take care of localizing the corresponding error message. Furthermore the framework will notify a firmware update failure if the update firmware operation takes longer than 30 minutes or if an unexpected exception occurred.
+
```

Comment:
It would be good to also comment on the cancel method. The handler interface provides a ```cancel()``` method, and the callback has a ```cancelled()``` method which I assume needs to be called once the transfer is finally cancelled?